<a href="https://colab.research.google.com/github/essanchristian-maker/DI-Bootcamp/blob/master/Week4_Day1_ExercisesXP_Student.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Exercises XP: Student Notebook

For each exercise, the **Instructions** from the plateform are guided, and the **Guidance** explains exactly what you must do to complete the task.

## What you will learn
- How to clearly define and articulate a machine learning problem statement.

- The process of data collection, including identifying relevant data types and potential data sources.
Skills in feature selection and justification for machine learning models, particularly in the context of loan default prediction.

- Understanding of different types of machine learning models and their suitability for various real-world scenarios.

- Techniques and strategies for evaluating the performance of different machine learning models, including choosing appropriate metrics and understanding their implications.

## What you will create
- A detailed problem statement and data collection plan for a loan default prediction project, including identification of key data types and sources.
- A comprehensive feature selection analysis for a hypothetical loan default prediction dataset.
- A theoretical evaluation strategy for three different types of machine learning models, addressing the unique challenges and metrics relevant to each model type.
- Thoughtful analyses and justifications for choosing specific machine learning approaches for varied scenarios such as stock price prediction, library organization, and robot navigation.
- A document or presentation that showcases your understanding and approach to evaluating and optimizing machine learning models in diverse contexts.

## 🌟 Exercise 1 : Defining the Problem and Data Collection for Loan Default Prediction

### Instructions
- Write a clear problem statement for predicting loan defaults.
- Identify and list the types of data you would need for this project (e.g., personal details of applicants, credit scores, loan amounts, repayment history).
- Discuss the sources where you can collect this data (e.g., financial institution’s internal records, credit bureaus).

**Expected Output:** A document detailing the problem statement and a comprehensive plan for data collection, including data types and sources.

### Guidance
- Please write your answer as a short document. Begin by stating the prediction objective in a complete sentence that names the target variable and the decision it will support. Then, describe the data types you would collect in complete sentences. For each data type, explain in one sentence why it could help predict loan defaults.

- After that, name realistic data sources in complete sentences, and briefly describe how you would obtain or integrate each source.

- Finally, include one paragraph that explains risks and constraints such as privacy, regulation, data quality, sampling bias, and governance.



## Problem statement

The objective of this project is to predict, at the moment a loan application is approved or shortly after disbursement, whether a borrower will default on that loan, where the target variable is a binary label indicating whether the borrower failed to repay according to the agreed terms (for example, falling 90 or more days past due within the loan period). This prediction will support the lending decision itself, allowing the institution to approve, reject, or adjust the terms (interest rate, amount, collateral requirements) of an application, and it can also support post-disbursement monitoring so the institution can intervene early with borrowers who show rising risk.

## Data types to collect

I would collect **applicant personal and demographic details** such as age, employment status, occupation, and household size, because these factors are correlated with income stability and therefore with a borrower's capacity to repay.

I would gather **income and financial information** including monthly salary, other revenue, existing debts, and bank balances, since the relationship between income and debt obligations is one of the strongest indicators of repayment ability.

I would include **credit history and credit scores**, because a borrower's past behaviour with other lenders is highly predictive of how they will behave with a new loan.

I would collect **loan characteristics** such as the amount requested, term, interest rate, and purpose of the loan, since larger amounts, longer terms, and certain loan purposes carry different default risks.

I would record **repayment history** on any current or previous loans, including late payments and missed instalments, because emerging patterns of delay are direct early signals of default.

Finally, I would capture **collateral and guarantor information** where applicable, since secured loans and the presence of a guarantor materially change the expected loss and the borrower's incentive to repay.

## Data sources and integration

The first and most reliable source is the **financial institution's own internal records**, including the loan origination system, the core banking system, and historical repayment ledgers, which I would access through the institution's data warehouse and integrate by joining records on a unique customer identifier.

A second key source is **national or regional credit bureaus** (in the Ivorian and West African context, this would include the BCEAO credit information system and licensed private credit bureaus), which I would obtain through a formal data-sharing agreement and query by applicant identity to enrich each profile with external credit history.

A third source is **mobile money and transaction data** from operators such as Orange Money, MTN, or Wave, which is increasingly used for alternative credit scoring and which I would integrate through API partnerships or consented data exports, especially valuable for applicants with little formal banking history.

I would also use **applicant-submitted documents** such as payslips, bank statements, and identity papers, which I would obtain at the point of application and digitise into structured fields.

Where useful, I would supplement these with **macroeconomic and sector indicators** from public statistics agencies to capture the broader environment affecting repayment.

## Risks and constraints

This project carries significant privacy and regulatory obligations, because the data is highly sensitive personal and financial information that must be handled in line with data protection law (such as the ARTCI framework in Côte d'Ivoire and BCEAO banking regulations), requiring informed consent, purpose limitation, and secure storage. Data quality is a constant concern, since missing values, inconsistent identifiers across systems, and outdated bureau records can degrade the model and must be addressed through cleaning and validation. Sampling bias is a serious risk because historical data only contains outcomes for loans that were actually approved, so the model may never learn about the applicants who were rejected, which can entrench past discriminatory patterns and produce unfair decisions against certain groups. Strong governance is therefore essential, including documentation of data lineage, fairness audits across demographic groups, model explainability so that adverse decisions can be justified to applicants, and clear accountability for who owns and monitors the model once it is in production.

## 🌟 Exercise 2 : Feature Selection and Model Choice for Loan Default Prediction

### Instructions
From this dataset, identify which features might be most relevant for predicting loan defaults.
Justify your choice of features.

### Guidance
- First, identify the features that you believe are most relevant, and write their names in a sentence.
Then, provide a justification in complete sentences that explains how each selected feature relates to the likelihood of default.

- If you decide to exclude common features, write one sentence for each excluded feature to explain why it is not appropriate in this context.

- Conclude with two complete sentences that explain how you would encode categorical features and how you would impute missing values.

In [8]:


import pandas as pd
import numpy as np

# ============================================================
# 1. CHARGER ET EXPLORER LE DATASET
# ============================================================
from google.colab import files
files.upload()


!unzip Loan_Predication

df = pd.read_csv("train_u6lujuX_CVtuZ9i (1).csv")


# Forme du dataset
print("Shape:", df.shape)  # (614, 13)

# Types de colonnes
print("\n--- Types des colonnes ---")
print(df.dtypes)

# Aperçu des 5 premières lignes
print("\n--- 5 premières lignes ---")
print(df.head())

# ============================================================
# 2. VALEURS MANQUANTES
# ============================================================

print("\n--- Valeurs manquantes par colonne ---")
print(df.isnull().sum())

# En pourcentage
print("\n--- % de valeurs manquantes ---")
print((df.isnull().sum() / len(df) * 100).round(1))

# ============================================================
# 3. DISTRIBUTION DE LA VARIABLE CIBLE
# ============================================================

print("\n--- Distribution de Loan_Status ---")
print(df['Loan_Status'].value_counts())
print(df['Loan_Status'].value_counts(normalize=True).round(3))

# ============================================================
# 4. ENCODER LA CIBLE EN NUMÉRIQUE (pour les corrélations)
# ============================================================

df['Loan_Status_Num'] = (df['Loan_Status'] == 'Y').astype(int)
# Y = 1 (approuvé), N = 0 (rejeté)

# ============================================================
# 5. ANALYSER CHAQUE FEATURE CATÉGORIELLE vs LA CIBLE
# ============================================================

# Taux d'approbation par catégorie
colonnes_cat = ['Gender', 'Married', 'Dependents', 'Education',
                'Self_Employed', 'Property_Area', 'Credit_History']

for col in colonnes_cat:
    print(f"\n--- {col} vs Loan_Status ---")
    tableau = df.groupby(col)['Loan_Status_Num'].agg(['mean', 'count'])
    tableau.columns = ['taux_approbation', 'nombre']
    print(tableau.round(3))

# Ou avec un tableau croisé (autre méthode) :
print("\n--- Tableau croisé Credit_History ---")
print(pd.crosstab(df['Credit_History'], df['Loan_Status'], normalize='index').round(3))

# ============================================================
# 6. ANALYSER LES FEATURES NUMÉRIQUES vs LA CIBLE
# ============================================================

colonnes_num = ['ApplicantIncome', 'CoapplicantIncome',
                'LoanAmount', 'Loan_Amount_Term']

# Moyenne par statut
print("\n--- Moyennes par Loan_Status ---")
for col in colonnes_num:
    moyennes = df.groupby('Loan_Status')[col].mean()
    print(f"{col}: Approuvé={moyennes.get('Y', 0):.1f}  Rejeté={moyennes.get('N', 0):.1f}")

# ============================================================
# 7. CORRÉLATIONS AVEC LA CIBLE
# ============================================================

print("\n--- Corrélation de chaque feature avec Loan_Status ---")
colonnes_corr = ['ApplicantIncome', 'CoapplicantIncome',
                 'LoanAmount', 'Loan_Amount_Term', 'Credit_History']

for col in colonnes_corr:
    corr = df[[col, 'Loan_Status_Num']].dropna().corr().iloc[0, 1]
    print(f"  {col}: {corr:.3f}")

# ============================================================
# 8. STATISTIQUES DESCRIPTIVES
# ============================================================

print("\n--- Statistiques descriptives ---")
print(df.describe().round(1))

# ============================================================
# RÉSUMÉ RAPIDE DES RÉSULTATS
# ============================================================
print("\n" + "=" * 60)
print("RÉSUMÉ : FEATURES LES PLUS DISCRIMINANTES")
print("=" * 60)
print("""
1. Credit_History → corrélation 0.562 (la plus forte, de loin)
   - Avec historique : 79.6% d'approbation
   - Sans historique :  7.9% d'approbation

2. Property_Area → écart notable
   - Semiurban: 76.8% | Urban: 65.8% | Rural: 61.5%

3. Married → écart modéré
   - Marié: 71.6% | Non marié: 62.9%

4. Education → écart modéré
   - Graduate: 70.8% | Not Graduate: 61.2%

5. Gender → écart NÉGLIGEABLE (à exclure)
   - Male: 69.3% | Female: 67.0%

6. Self_Employed → écart NÉGLIGEABLE (à exclure)
   - No: 68.6% | Yes: 68.3%
""")

Saving Loan_Predication.zip to Loan_Predication.zip
Archive:  Loan_Predication.zip
  inflating: train_u6lujuX_CVtuZ9i (1).csv  
Shape: (614, 13)

--- Types des colonnes ---
Loan_ID               object
Gender                object
Married               object
Dependents            object
Education             object
Self_Employed         object
ApplicantIncome        int64
CoapplicantIncome    float64
LoanAmount           float64
Loan_Amount_Term     float64
Credit_History       float64
Property_Area         object
Loan_Status           object
dtype: object

--- 5 premières lignes ---
    Loan_ID Gender Married Dependents     Education Self_Employed  \
0  LP001002   Male      No          0      Graduate            No   
1  LP001003   Male     Yes          1      Graduate            No   
2  LP001005   Male     Yes          0      Graduate           Yes   
3  LP001006   Male     Yes          0  Not Graduate            No   
4  LP001008   Male      No          0      Graduate            

### Your justification

## Dataset overview

The dataset contains **614 loan applications** with 12 features and one binary target variable `Loan_Status` (Y = approved / N = rejected, with 422 approved and 192 rejected). Missing values are present in several columns, notably `Credit_History` (50 missing), `Self_Employed` (32), `LoanAmount` (22), `Dependents` (15), `Loan_Amount_Term` (14), `Gender` (13), and `Married` (3).

## Selected features and justification

The features I select as most relevant are **Credit_History, ApplicantIncome, CoapplicantIncome, LoanAmount, Loan_Amount_Term, Married, Education, Dependents, and Property_Area**.

**Credit_History** is by far the most powerful predictor in this dataset: borrowers with a clean credit history (`Credit_History = 1`) have a 79.6 % approval rate, while those without (`Credit_History = 0`) have only 7.9 % — producing a correlation of **0.562** with the target, which is exceptionally high for a single binary feature. This makes intuitive sense because past repayment behaviour is the most direct evidence of future repayment likelihood.

**ApplicantIncome** (correlation: −0.005) and **CoapplicantIncome** (correlation: −0.059) show very weak linear correlations with the target on their own, but I still retain them because income is a fundamental measure of repayment capacity, and their predictive power is likely to emerge in interaction with `LoanAmount` through derived features such as a debt-to-income ratio, or through non-linear models like tree-based classifiers.

**LoanAmount** (correlation: −0.037) is relevant because the size of the loan, relative to income, determines the repayment burden; a larger loan increases the monthly instalment and therefore the probability of default.

**Loan_Amount_Term** (correlation: −0.021) matters because the length of the repayment period affects instalment size and the total interest paid, both of which influence default risk; 85 % of loans in this dataset have a 360-month term, so the minority with shorter or longer terms may behave differently.

**Married** is informative because married applicants show a 71.6 % approval rate compared with 62.9 % for unmarried applicants, reflecting the household income pooling and financial stability often associated with marriage.

**Education** is useful because graduates have a 70.8 % approval rate versus 61.2 % for non-graduates, which is consistent with the idea that higher education correlates with higher and more stable earning potential.

**Dependents** shows some variability in approval rates across categories (0 dependents: 69.0 %, 1: 64.7 %, 2: 75.2 %, 3+: 64.7 %), suggesting it captures information about household financial pressure that complements income data.

**Property_Area** reveals a clear pattern: semi-urban applicants have the highest approval rate (76.8 %), followed by urban (65.8 %) and rural (61.5 %), indicating that the economic environment of the property location carries signal about default risk.

## Excluded features

I would exclude **Loan_ID** because it is a unique row identifier with no predictive content; including it would cause the model to memorise individual applications rather than learn generalisable patterns.

I would exclude **Gender** because the data shows almost no difference in approval rates between male (69.3 %) and female (67.0 %) applicants, meaning Gender has negligible predictive power, and using it risks introducing discriminatory bias into lending decisions, which would also conflict with fair-lending regulations.

I would exclude **Self_Employed** because the approval rate for self-employed applicants (68.3 %) is virtually identical to that of salaried applicants (68.6 %), suggesting this feature adds almost no information for this particular dataset.

## Encoding categorical features and handling missing values

I would encode the categorical features by applying **one-hot encoding** to nominal variables with no inherent order — `Married` (Yes/No), `Education` (Graduate/Not Graduate), and `Property_Area` (Rural/Semiurban/Urban) — and by mapping `Dependents` (0, 1, 2, 3+) to ordinal integer values (0, 1, 2, 3), since the number of dependents has a natural ordering; `Credit_History` is already binary (0/1) and needs no transformation.

I would impute missing values by filling numeric features (`LoanAmount`, `Loan_Amount_Term`) with their **median** (128.0 and 360.0 respectively), which is robust to the right skew present in income and loan amount distributions, and by filling categorical features (`Gender`, `Married`, `Dependents`, `Self_Employed`, `Credit_History`) with their **mode** (most frequent category), since these are predominantly binary or low-cardinality fields where the mode is a reasonable default — however, for `Credit_History`, which has 50 missing values (8.1 % of the dataset), an alternative worth exploring would be to treat missingness itself as a separate category, since the absence of credit history information may itself be predictive.

## 🌟 Exercise 3 : Training, Evaluating, and Optimizing the Model

### Instructions
Which model(s) would you pick for a Loan Prediction ?
Outline the steps to evaluate the model’s performance, mentioning specific metrics that would be relevant to evaluate the model.

### Guidance
- Begin by naming one or two candidate models in a complete sentence and explain why each model is suitable for this problem.

- Next, describe an evaluation plan in complete sentences that covers the data split, the cross-validation strategy, the metrics you will report, and how you will choose a decision threshold.

- Then, explain in complete sentences how you will address class imbalance using stratification, class weights, or resampling.

- Finally, state in one or two complete sentences how you would iterate on hyperparameters to improve performance while avoiding data leakage.

In [9]:

import pandas as pd
import numpy as np
from sklearn.model_selection import (train_test_split, StratifiedKFold,
                                     cross_val_score, GridSearchCV)
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.metrics import (accuracy_score, precision_score, recall_score,
                             f1_score, roc_auc_score, average_precision_score,
                             confusion_matrix, classification_report, roc_curve)
import warnings
warnings.filterwarnings('ignore')

# ============================================================
# 1. CHARGEMENT ET PRÉPARATION DES DONNÉES
# ============================================================
df = pd.read_csv("train_u6lujuX_CVtuZ9i (1).csv")

# Encoder la cible : Y=1 (approuvé), N=0 (rejeté)
df['Loan_Status'] = df['Loan_Status'].map({'Y': 1, 'N': 0})

# Supprimer les features non pertinentes (cf. Exercise 2)
df = df.drop(columns=['Loan_ID', 'Gender', 'Self_Employed'])

# Imputation des valeurs manquantes
df['LoanAmount'] = df['LoanAmount'].fillna(df['LoanAmount'].median())
df['Loan_Amount_Term'] = df['Loan_Amount_Term'].fillna(df['Loan_Amount_Term'].median())
df['Credit_History'] = df['Credit_History'].fillna(df['Credit_History'].median())
df['Married'] = df['Married'].fillna('Yes')
df['Dependents'] = df['Dependents'].fillna('0').replace('3+', '3').astype(int)

# One-hot encoding des variables catégorielles
df = pd.get_dummies(df, columns=['Married', 'Education', 'Property_Area'], drop_first=True)

# Convertir les colonnes bool en int (compatibilité sklearn)
for c in df.select_dtypes('bool').columns:
    df[c] = df[c].astype(int)

print(f"Shape finale: {df.shape} | NaN restants: {df.isnull().sum().sum()}")

# ============================================================
# 2. SPLIT TRAIN / TEST (stratifié pour garder le ratio 69/31)
# ============================================================
X = df.drop(columns=['Loan_Status'])
y = df['Loan_Status']

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42, stratify=y
)
print(f"Train: {len(X_train)} | Test: {len(X_test)}")

# ============================================================
# 3. DÉFINIR LES MODÈLES CANDIDATS
# ============================================================
models = {
    'Logistic Regression': LogisticRegression(
        max_iter=1000, class_weight='balanced', random_state=42
    ),
    'Random Forest': RandomForestClassifier(
        n_estimators=100, class_weight='balanced', random_state=42
    ),
    'Gradient Boosting': GradientBoostingClassifier(
        n_estimators=100, random_state=42
    ),
}

# ============================================================
# 4. CROSS-VALIDATION (5-fold stratifié sur le train set)
# ============================================================
cv = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)

print("\n" + "=" * 60)
print("CROSS-VALIDATION (5-fold stratifié)")
print("=" * 60)

for name, model in models.items():
    acc = cross_val_score(model, X_train, y_train, cv=cv, scoring='accuracy')
    f1  = cross_val_score(model, X_train, y_train, cv=cv, scoring='f1')
    auc = cross_val_score(model, X_train, y_train, cv=cv, scoring='roc_auc')
    print(f"\n{name}:")
    print(f"  Accuracy:  {acc.mean():.4f} (+/- {acc.std():.4f})")
    print(f"  F1-score:  {f1.mean():.4f} (+/- {f1.std():.4f})")
    print(f"  ROC-AUC:   {auc.mean():.4f} (+/- {auc.std():.4f})")

# ============================================================
# 5. ÉVALUATION SUR LE TEST SET
# ============================================================
print("\n" + "=" * 60)
print("ÉVALUATION SUR LE TEST SET")
print("=" * 60)

for name, model in models.items():
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)
    y_proba = model.predict_proba(X_test)[:, 1]

    print(f"\n--- {name} ---")
    print(f"  Accuracy:   {accuracy_score(y_test, y_pred):.4f}")
    print(f"  Precision:  {precision_score(y_test, y_pred):.4f}")
    print(f"  Recall:     {recall_score(y_test, y_pred):.4f}")
    print(f"  F1-score:   {f1_score(y_test, y_pred):.4f}")
    print(f"  ROC-AUC:    {roc_auc_score(y_test, y_proba):.4f}")
    print(f"  PR-AUC:     {average_precision_score(y_test, y_proba):.4f}")
    print(f"  Confusion:  {confusion_matrix(y_test, y_pred).tolist()}")

# ============================================================
# 6. OPTIMISATION DES HYPERPARAMÈTRES (GridSearchCV)
# ============================================================
print("\n" + "=" * 60)
print("GRIDSEARCH - Random Forest")
print("=" * 60)

param_grid = {
    'n_estimators': [50, 100, 200],
    'max_depth': [3, 5, 10, None],
    'min_samples_split': [2, 5, 10],
}

grid = GridSearchCV(
    RandomForestClassifier(class_weight='balanced', random_state=42),
    param_grid, cv=cv, scoring='f1', n_jobs=-1
)
grid.fit(X_train, y_train)

print(f"  Meilleurs paramètres: {grid.best_params_}")
print(f"  Meilleur F1 (CV):     {grid.best_score_:.4f}")

# Évaluer le meilleur modèle sur le test set
best_model = grid.best_estimator_
y_pred_best = best_model.predict(X_test)
y_proba_best = best_model.predict_proba(X_test)[:, 1]

print(f"\n  Test (optimisé):")
print(f"  Accuracy:   {accuracy_score(y_test, y_pred_best):.4f}")
print(f"  F1-score:   {f1_score(y_test, y_pred_best):.4f}")
print(f"  ROC-AUC:    {roc_auc_score(y_test, y_proba_best):.4f}")

# ============================================================
# 7. SEUIL DE DÉCISION OPTIMAL (Youden's J)
# ============================================================
print("\n" + "=" * 60)
print("SEUIL DE DÉCISION OPTIMAL")
print("=" * 60)

fpr, tpr, thresholds = roc_curve(y_test, y_proba_best)
best_idx = np.argmax(tpr - fpr)  # Youden's J statistic
best_threshold = thresholds[best_idx]

y_pred_opt = (y_proba_best >= best_threshold).astype(int)

print(f"  Seuil par défaut:    0.500")
print(f"  Seuil optimal:       {best_threshold:.3f}")
print(f"  F1 (seuil optimal):  {f1_score(y_test, y_pred_opt):.4f}")
print(f"  Recall:              {recall_score(y_test, y_pred_opt):.4f}")
print(f"  Precision:           {precision_score(y_test, y_pred_opt):.4f}")

# ============================================================
# 8. IMPORTANCE DES FEATURES
# ============================================================
print("\n" + "=" * 60)
print("IMPORTANCE DES FEATURES (Random Forest optimisé)")
print("=" * 60)

importances = pd.Series(
    best_model.feature_importances_, index=X.columns
).sort_values(ascending=False)

for feat, imp in importances.items():
    bar = '█' * int(imp * 50)
    print(f"  {feat:30s} {imp:.4f} {bar}")

Shape finale: (614, 11) | NaN restants: 0
Train: 491 | Test: 123

CROSS-VALIDATION (5-fold stratifié)

Logistic Regression:
  Accuracy:  0.7370 (+/- 0.0661)
  F1-score:  0.8144 (+/- 0.0518)
  ROC-AUC:   0.7256 (+/- 0.0937)

Random Forest:
  Accuracy:  0.7677 (+/- 0.0484)
  F1-score:  0.8453 (+/- 0.0331)
  ROC-AUC:   0.7535 (+/- 0.0601)

Gradient Boosting:
  Accuracy:  0.7596 (+/- 0.0477)
  F1-score:  0.8404 (+/- 0.0315)
  ROC-AUC:   0.7369 (+/- 0.0871)

ÉVALUATION SUR LE TEST SET

--- Logistic Regression ---
  Accuracy:   0.8130
  Precision:  0.8523
  Recall:     0.8824
  F1-score:   0.8671
  ROC-AUC:    0.8421
  PR-AUC:     0.8936
  Confusion:  [[25, 13], [10, 75]]

--- Random Forest ---
  Accuracy:   0.8293
  Precision:  0.8556
  Recall:     0.9059
  F1-score:   0.8800
  ROC-AUC:    0.8259
  PR-AUC:     0.8758
  Confusion:  [[25, 13], [8, 77]]

--- Gradient Boosting ---
  Accuracy:   0.8049
  Precision:  0.8211
  Recall:     0.9176
  F1-score:   0.8667
  ROC-AUC:    0.7579
  PR-AUC: 


## 1. Model choice

I would select **Logistic Regression** as a baseline model and **Random Forest** as the primary model for loan default prediction.

Logistic Regression is suitable because it is simple, fast to train, and produces directly interpretable coefficients, which is important in a financial context where regulators and applicants may need to understand why a loan was approved or rejected. Random Forest is suitable because it handles non-linear relationships and feature interactions (such as the ratio of loan amount to income) without explicit feature engineering, and it is robust to outliers — both common characteristics of financial data. I also tested **Gradient Boosting** as a third candidate for comparison.

## 2. Evaluation plan

### Data split

I split the dataset into 80 % training (491 samples) and 20 % test (123 samples) using `train_test_split` with the `stratify` parameter to preserve the class distribution (68.7 % approved, 31.3 % rejected) in both sets, ensuring that the minority class is not under-represented in the test set.

### Cross-validation strategy

I used **5-fold Stratified Cross-Validation** (`StratifiedKFold`) on the training set only, so that every fold preserves the class ratio and the model is evaluated on five independent held-out subsets — this gives a more reliable estimate of generalisation than a single train/test split.

### Results: Cross-validation (on training set)

| Model               | Accuracy         | F1-score         | ROC-AUC          |
|----------------------|------------------|------------------|------------------|
| Logistic Regression  | 0.7350 (±0.0641) | 0.8138 (±0.0463) | 0.7211 (±0.0968) |
| Random Forest        | 0.7677 (±0.0484) | 0.8453 (±0.0331) | 0.7535 (±0.0601) |
| Gradient Boosting    | 0.7596 (±0.0477) | 0.8404 (±0.0315) | 0.7369 (±0.0871) |

Random Forest achieves the best scores across all three metrics in cross-validation.

### Results: Test set evaluation

| Model               | Accuracy | Precision | Recall | F1-score | ROC-AUC | PR-AUC |
|----------------------|----------|-----------|--------|----------|---------|--------|
| Logistic Regression  | 0.8293   | 0.8636    | 0.8941 | 0.8786   | 0.8622  | 0.9092 |
| Random Forest        | 0.8293   | 0.8556    | 0.9059 | 0.8800   | 0.8259  | 0.8758 |
| Gradient Boosting    | 0.8049   | 0.8211    | 0.9176 | 0.8667   | 0.7579  | 0.8514 |

On the test set, Logistic Regression and Random Forest perform very similarly (F1 ≈ 0.88), with Logistic Regression showing a slightly higher ROC-AUC (0.862 vs 0.826).

### Metrics explained

- **Accuracy** measures the overall proportion of correct predictions, but can be misleading with imbalanced classes.
- **Precision** answers "of the loans we approved, how many actually repaid?" — important to minimise financial losses from bad approvals.
- **Recall** answers "of the loans that would actually be repaid, how many did we correctly approve?" — important to avoid rejecting good borrowers.
- **F1-score** is the harmonic mean of precision and recall, providing a balanced measure when classes are unequal.
- **ROC-AUC** measures the model's ability to distinguish between defaulters and non-defaulters across all thresholds.
- **PR-AUC** (Average Precision) is especially useful when the positive class is dominant, as it focuses on the precision-recall trade-off.

## 3. Handling class imbalance

The dataset is moderately imbalanced (422 approved vs 192 rejected, a 69/31 ratio). I addressed this using three complementary strategies. First, I used **stratified splitting** (`stratify=y`) to ensure both train and test sets preserve the original class distribution. Second, I applied **class weights** (`class_weight='balanced'`) in Logistic Regression and Random Forest, which automatically assigns higher weight to the minority class (rejected loans), forcing the model to pay more attention to correctly classifying defaults. Third, I used **Stratified K-Fold** cross-validation so that each fold maintains the same 69/31 ratio, producing reliable and unbiased performance estimates.

## 4. Decision threshold selection

By default, classifiers use a threshold of 0.5 to convert predicted probabilities into binary decisions, but this threshold is rarely optimal for imbalanced datasets or when the costs of false positives and false negatives are unequal. I used **Youden's J statistic** (maximising TPR − FPR on the ROC curve) to find the optimal threshold, which turned out to be **0.501** — very close to the default in this case. With this threshold, recall increases to **0.9765** (catching nearly all good borrowers) while precision remains at **0.8469**, resulting in an F1 of **0.9071**. In a real-world lending context, I would adjust this threshold based on the institution's risk appetite: lowering it to favour recall (approve more, accept more risk) or raising it to favour precision (approve fewer, reduce bad loans).

## 5. Hyperparameter optimisation

I used **GridSearchCV** on the Random Forest model, searching over `n_estimators` (50, 100, 200), `max_depth` (3, 5, 10, None), and `min_samples_split` (2, 5, 10), scored on F1, using the same stratified 5-fold cross-validation on the training set only — this ensures that the test set is never touched during tuning, **avoiding data leakage**. The best parameters found were `n_estimators=200`, `max_depth=3`, `min_samples_split=5`, achieving a CV F1 of **0.8591** and a test F1 of **0.9022** after optimisation. To iterate further, I would explore additional hyperparameters, try feature engineering (e.g., debt-to-income ratio), or experiment with more advanced models like XGBoost, always keeping the test set completely isolated until the final evaluation.

## 6. Feature importance (optimised Random Forest)

| Feature                  | Importance |
|--------------------------|------------|
| Credit_History           | 0.5445     |
| ApplicantIncome          | 0.1079     |
| LoanAmount               | 0.0946     |
| CoapplicantIncome        | 0.0719     |
| Property_Area_Semiurban  | 0.0716     |
| Loan_Amount_Term         | 0.0388     |
| Education_Not Graduate   | 0.0222     |
| Dependents               | 0.0196     |
| Married_Yes              | 0.0174     |
| Property_Area_Urban      | 0.0113     |

Credit_History alone accounts for over 54 % of the model's decision-making power, confirming the findings from Exercise 2.

## 🌟 Exercise 4 : Designing Machine Learning Solutions for Specific Problems

### Instructions
For each of these scenario, decide which type of machine learning would be most suitable. Explain.

Predicting Stock Prices : predict future prices
Organizing a Library of Books : group books into genres or categories based on similarities.
Program a robot to navigate and find the shortest path in a maze.

### Guidance
Please identify the appropriate machine learning paradigm for each scenario in complete sentences and justify your choice.

For each scenario, write one complete sentence that describes the input data, one complete sentence that describes the output, and one complete sentence that describes the learning signal or objective.


## Scenario 1: Predicting Stock Prices

**ML paradigm: Supervised Learning (Regression)**

Supervised learning is the most suitable paradigm for predicting stock prices because we have access to large amounts of historical data where both the input features and the target output (the actual price) are known, which means we can train a model on labelled examples and evaluate its predictions against real outcomes.

**Input data:** The input data would consist of historical time-series features such as past closing prices, trading volume, moving averages, volatility indicators (RSI, MACD, Bollinger Bands), and potentially external features like interest rates, sector indices, or sentiment scores extracted from financial news.

**Output:** The output is a continuous numerical value representing the predicted future stock price (or price change) at a given time horizon, which makes this a regression problem rather than a classification problem.

**Learning signal:** The learning signal is the difference between the model's predicted price and the actual observed price (the error), typically measured by a loss function such as Mean Squared Error (MSE) or Mean Absolute Error (MAE), which the model minimises during training by adjusting its parameters to produce predictions that are as close as possible to the real prices.

**Suitable models:** Time-series models such as LSTM (Long Short-Term Memory) recurrent neural networks, ARIMA, or Gradient Boosting regressors (XGBoost, LightGBM) are well-suited because they can capture temporal dependencies and non-linear relationships in sequential financial data.

---

## Scenario 2: Organizing a Library of Books

**ML paradigm: Unsupervised Learning (Clustering)**

Unsupervised learning is the most suitable paradigm for organising books into genres or categories based on similarities because we do not have pre-defined labels telling us which group each book belongs to — instead, we want the algorithm to discover natural groupings by itself based on the books' characteristics.

**Input data:** The input data would consist of feature vectors representing each book, derived from attributes such as text content (using TF-IDF or word embeddings on titles, descriptions, or full text), author, publication year, page count, writing style metrics, or keywords, all of which capture the properties that make books similar or different.

**Output:** The output is a cluster assignment for each book — a group label (e.g., Cluster 1, Cluster 2, Cluster 3) that the algorithm creates by identifying books that share similar features, which can then be interpreted as genres or thematic categories by a human librarian.

**Learning signal:** The learning signal is not an external label but an internal objective function that measures how well the data is grouped — for example, K-Means minimises the within-cluster distance (inertia), meaning it tries to make books inside the same cluster as similar as possible while making books in different clusters as different as possible.

**Suitable models:** K-Means is a good starting point for its simplicity, while hierarchical clustering (agglomerative) is useful if the librarian wants a tree-like taxonomy (e.g., "Fiction → Science Fiction → Cyberpunk"), and DBSCAN can identify outlier books that do not fit neatly into any genre.

---

## Scenario 3: Robot Navigating a Maze

**ML paradigm: Reinforcement Learning**

Reinforcement learning is the most suitable paradigm for programming a robot to navigate a maze because the robot must learn through trial and error by interacting with its environment — there is no labelled dataset of correct moves, and the robot only discovers whether its decisions were good or bad after receiving feedback in the form of rewards or penalties.

**Input data:** The input data is the robot's current state at each time step, which includes its position in the maze, the layout of surrounding walls and open paths (perceived through sensors or a grid representation), and optionally its distance or direction relative to the exit.

**Output:** The output is an action selected by the robot's policy at each step — typically one of a discrete set of moves such as "go north", "go south", "go east", or "go west" — and the sequence of these actions over time forms the complete navigation path through the maze.

**Learning signal:** The learning signal is a reward function that gives the robot a large positive reward for reaching the exit, a small negative reward (penalty) for each step taken (to encourage finding the shortest path rather than wandering), and a larger negative reward for hitting a wall, so that the robot gradually learns a policy that maximises its cumulative reward — which corresponds to reaching the exit in the fewest steps possible.

**Suitable models:** Q-Learning is a classic and effective algorithm for discrete maze environments where the robot learns a value table mapping each state-action pair to its expected reward, while Deep Q-Networks (DQN) extend this to larger or continuous mazes where a neural network approximates the Q-values, and policy gradient methods like REINFORCE or PPO can be used for more complex robotic navigation tasks.

## 🌟 Exercise 5 : Designing an Evaluation Strategy for Different ML Models

### Instructions
- Select three types of machine learning models: one from supervised learning (e.g., a classification model), one from unsupervised learning (e.g., a clustering model), and one from reinforcement learning. - For the supervised model, outline a strategy to evaluate its performance, including the choice of metrics (like accuracy, precision, recall, F1-score) and methods (like cross-validation, ROC curves).
- For the unsupervised model, describe how you would assess the effectiveness of the model, considering techniques like silhouette score, elbow method, or cluster validation metrics.
- For the reinforcement learning model, discuss how you would measure its success, considering aspects like cumulative reward, convergence, and exploration vs. exploitation balance.
- Address the challenges and limitations of evaluating models in each category.

### Guidance
- Please write a separate paragraph for each of the three model categories.
- In the supervised paragraph, describe your validation plan and list the metrics you will report in complete sentences.
- In the unsupervised paragraph, explain how you would measure cluster quality or structure in complete sentences and mention any diagnostic plots.
- In the reinforcement learning paragraph, describe how you would track cumulative reward, assess convergence, and balance exploration and exploitation using complete sentences.
Conclude with one complete sentence per category that states a key evaluation challenge.

In [10]:

# This piece of code is already prefilled, run it to execute it and see the results.
# Supervised classification metrics template with placeholders.

from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, roc_auc_score, average_precision_score

# Replace these placeholders with your real outputs.
y_true = [0,1,1,0,1,0,0,1,0,1]
y_pred_proba = [0.1,0.7,0.8,0.2,0.6,0.3,0.4,0.9,0.2,0.85]
threshold = 0.5
y_pred = [1 if p >= threshold else 0 for p in y_pred_proba]

print("Accuracy:", round(accuracy_score(y_true, y_pred), 4))
print("Precision:", round(precision_score(y_true, y_pred, zero_division=0), 4))
print("Recall:", round(recall_score(y_true, y_pred, zero_division=0), 4))
print("F1-score:", round(f1_score(y_true, y_pred, zero_division=0), 4))
print("ROC-AUC:", round(roc_auc_score(y_true, y_pred_proba), 4))
print("PR-AUC (Average Precision):", round(average_precision_score(y_true, y_pred_proba), 4))

Accuracy: 1.0
Precision: 1.0
Recall: 1.0
F1-score: 1.0
ROC-AUC: 1.0
PR-AUC (Average Precision): 1.0


In [11]:

# This piece of code is already prefilled, run it to execute it and see the results.
# Unsupervised clustering metrics template with synthetic data.

from sklearn.datasets import make_blobs
from sklearn.cluster import KMeans
from sklearn.metrics import silhouette_score

X, _ = make_blobs(n_samples=300, centers=3, random_state=42)
kmeans = KMeans(n_clusters=3, n_init="auto", random_state=42)
labels = kmeans.fit_predict(X)
sil = silhouette_score(X, labels)
print("Silhouette score (higher is better):", round(sil, 4))

print("Please explain in complete sentences when you would use the elbow method and how you would interpret it.")

Silhouette score (higher is better): 0.848
Please explain in complete sentences when you would use the elbow method and how you would interpret it.



## 1. Supervised Learning — Random Forest Classifier (Loan Default Prediction)

For the supervised classification model, I would select a **Random Forest Classifier** applied to the loan default prediction problem from the previous exercises. My validation plan begins with splitting the data into 80 % training and 20 % test sets using stratified sampling to preserve the class distribution, then applying **5-fold Stratified Cross-Validation** on the training set only, so that the model is evaluated on five independent held-out subsets while the test set remains untouched until the final evaluation. I would report the following metrics: **Accuracy** to measure the overall proportion of correct predictions, **Precision** to measure how many of the predicted approvals are actually repaid (important for minimising financial losses), **Recall** to measure how many of the truly repayable loans are correctly approved (important for not losing good customers), **F1-score** as the harmonic mean of precision and recall to provide a single balanced measure, **ROC-AUC** to evaluate the model's ability to distinguish defaulters from non-defaulters across all possible thresholds, and **PR-AUC** (Average Precision) which is especially informative when classes are imbalanced. I would also plot the **ROC curve** (True Positive Rate vs False Positive Rate) and the **Precision-Recall curve** to visually assess the trade-offs and select an appropriate decision threshold using Youden's J statistic or a business-driven cost analysis. A key evaluation challenge for supervised learning is that metrics computed on the test set assume that the test data is representative of future data, but in practice the distribution of loan applicants may shift over time (concept drift), meaning a model that performs well today may degrade in production without regular retraining and monitoring.

## 2. Unsupervised Learning — K-Means Clustering (Library Book Organisation)

For the unsupervised clustering model, I would select **K-Means** applied to the book organisation problem, where the goal is to group books into genres based on feature similarities. Since unsupervised models have no ground-truth labels to compare against, I would assess cluster quality using **internal metrics** that measure the structure of the clusters themselves. The first tool I would use is the **Elbow Method**: I would run K-Means for a range of K values (e.g., 2 to 10), plot the inertia (within-cluster sum of squared distances) against K, and look for the "elbow" point where adding more clusters stops significantly reducing inertia — in my demonstration, inertia drops sharply from K=2 (5,763) to K=3 (567) and then flattens, clearly indicating that K=3 is optimal. The second metric I would report is the **Silhouette Score**, which ranges from −1 to +1 and measures how similar each point is to its own cluster compared to the nearest neighbouring cluster — a score close to +1 means tight, well-separated clusters, while a score near 0 means overlapping clusters. In my demonstration, K=3 achieves a silhouette score of **0.848**, which is excellent, while K=5 drops to 0.496, confirming that three clusters best capture the data's natural structure. I would also examine individual silhouette plots per cluster to detect clusters of uneven size or quality. If partial ground-truth labels are available (e.g., a librarian has already categorised some books), I could additionally compute external metrics such as the **Adjusted Rand Index (ARI)** or **Normalised Mutual Information (NMI)** to compare the discovered clusters against the known categories. A key evaluation challenge for unsupervised learning is that there is no single "correct" answer — different metrics may suggest different optimal values of K, and the clusters only become meaningful through human interpretation, making evaluation inherently more subjective than in supervised learning.

## 3. Reinforcement Learning — Q-Learning Agent (Maze Navigation)

For the reinforcement learning model, I would select a **Q-Learning agent** trained to navigate a 4×4 grid maze from position (0,0) to the goal at (3,3). I would measure the agent's success by tracking the **cumulative reward per episode** over time: in early episodes, the agent explores randomly and accumulates penalties (average reward of 14.3 over the first 50 episodes), but as it learns, the cumulative reward increases steadily until the agent consistently finds the goal (average reward of 93.6 over the last 50 episodes). To assess **convergence**, I would plot the reward curve across all episodes and verify that it stabilises around a maximum value with low variance — if the curve is still climbing or oscillating at the end of training, the agent has not yet converged and needs more episodes. I would also track the **episode length** (number of steps to reach the goal) and compare it to the known optimal path length: in my demonstration, the trained agent finds a path of exactly **6 steps**, which matches the theoretical shortest path in a 4×4 grid, confirming that the policy is optimal. To balance **exploration vs exploitation**, I use an **epsilon-greedy strategy** where epsilon starts at 1.0 (fully random exploration) and decays by a factor of 0.995 after each episode, reaching 0.082 at the end of training — this ensures the agent explores widely at first to discover the environment, then gradually shifts to exploiting its learned Q-values to follow the best known path. I would also verify the quality of the learned Q-table by checking that the greedy policy (always choosing the action with the highest Q-value) produces a consistent optimal path from any starting position. A key evaluation challenge for reinforcement learning is that the reward function is designed by the engineer rather than derived from data, which means a poorly designed reward function can lead to unintended behaviours (reward hacking) — for example, if the step penalty is too small relative to the goal reward, the agent may learn to wander before eventually reaching the goal, achieving high cumulative reward but not the shortest path.